wiki/Bazel-User-Guide.md

```diff
@@ -0,0 +1,277 @@
+**[Packages](https://github.com/oppia/oppia-android/tree/develop/app)**
+
+A package is defined as a directory containing a file named BUILD or BUILD.bazel.
+
+**[Binary rules](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/BUILD.bazel#L3)**
+
+A rule specifies the relationship between inputs and outputs, and the steps to build the outputs.
+In Android, rules are defined using `android_binary`. Android rules for testing are `android_instrumentation_test` and `android_local_test`.
+
+**[BUILD files](https://github.com/oppia/oppia-android/blob/develop/app/BUILD.bazel)**
+
+Every package contains a BUILD file. This file is written in Starlark Language. In this Build file for module-level, we generally define `android_library`, `kt_android_library` to build our package files as per the requirement.
+
+**[Dependencies](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/BUILD.bazel#L16)**
+
+A target A depends upon a target B if B is needed by A at build. `A -> B`
+
+```
+deps = [ "//app",]
+```
+Here, `deps` is used to define the dependencies which is a type of dependencies called `deps dependencies` and it includes the files/directory/target which are dependent. From the above example the dependency is the `app` target which is defined in the [Build file of app package](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L616).
+
+Example of Dependencies
+1. [srcs dependencies](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L617)
+2. [deps dependencies](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L622)
+
+**[Loading an extension](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L13)**
+
+Bazel extensions are files ending in .bzl. Use the load statement to import a symbol from an extension.
+
+```
+load("@io_bazel_rules_kotlin//kotlin:android.bzl", "kt_android_library")
+```
+Here, we are loading `android.bzl` and we are going to use it with a symbol name `kt_android_library`.
+Arguments to the load function must be string literals. load statements must appear at top-level in the file.
+
+**[Visibility of a file target](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L621)**
+
+With the example from our codebase, target `app` whose visibility is public. 
+
+ - `visibility = ["//visibility:public"],` - Anyone can use this target.
+ - `"//visibility:private"` - Only targets in this package can use this target.
+
+**[Testing](https://github.com/oppia/oppia-android/blob/ba8d914480251e4a8543feb63a93b6c91e0a5a2f/app/BUILD.bazel#L719)**
```

Comment:
Can we update this to point to an oppia_android_test now that app_test was removed? I missed it in my PR since there wasn't an easy way to find this reference.

Reply:
Done